### PR TITLE
PorousFlow physics may now be block restricted

### LINKED ISSUE #20260: Allow PorousFlow to be entirely block restricted
## Reason

At the moment, in a PorousFlow simulation, the porous-flow physics must be active over the entire mesh.  This is because of the `PorousFlowAddMaterialJoiner` `Action` that adds new `Materials` that are not block-restricted.  We would like to build a model in which porous flow occurs on part of the mesh only, and some other physics governs the other part.

## Design

Possibly the simplest design is to include a `blocks` input parameter in the `PorousFlowDictator`.  It defaults to "nothing", meaning that PorousFlow occurs over the entire mesh - that means no current input files need re-writing.  Then `PorousFlowAddMaterialJoiner` can query this input parameter to see whether the Joiner `Materials` that it adds need to be block restricted.

@cpgr can comment

## Impact

Add ability to have different physics on different parts of the mesh

### PR DESCRIPTION
**Pull request by WilkAndy**
Fixes #20260

### CODE REVIEW

File: modules/porous_flow/src/actions/PorousFlowAddMaterialJoiner.C [hunk 1]
```diff
@@ -61,6 +61,36 @@ PorousFlowAddMaterialJoiner::act()
...
+              const std::vector<SubdomainName> & bl =
+                  params2.get<std::vector<SubdomainName>>("block");
+              _blocks.insert(_blocks.end(), bl.begin(), bl.end());
+            }
+          }
+          // eliminate any repeated entries in _blocks
+          const std::set<SubdomainName> unique(_blocks.begin(), _blocks.end());
+          _blocks.assign(unique.begin(), unique.end());
+        }
+
```
Comment (cpgr) [inline modules/porous_flow/src/actions/PorousFlowAddMaterialJoiner.C:93]: Could `_blocks` be a set - I don't know? Perhaps you can't insert a vector `bl` easily though? I really don't know.

**Review by cpgr — DISMISSED** [review]
What a bug! I never envisaged this occurring when I wrote this action. Thanks for fixing this up.

File: modules/porous_flow/src/actions/PorousFlowAddMaterialJoiner.C (hunk 1 above)
Comment (WilkAndy) [reply modules/porous_flow/src/actions/PorousFlowAddMaterialJoiner.C:91]: done

**Comment by WilkAndy** [comment]
Requesting review from @loganharbour (or some other INL person).

**Comment by garciapintado** [comment]
Hi @WilkAndy  and @cpgr,
Many thanks for this!! This is great!
Looking at the new example (multiblock.i) file you’ve added (Andy), and thinking about PorousFlowMaterialJoiner, I am wondering if we can safely assume that variables temp (in blocks 0 1; controlled by PorousFlow) and temp1 (in block 2; controlled by a non-PorousFlow Diffusion Kernel) in the example will be able to be coupled through their common bounding side?

**Comment by loganharbour** [comment]
@WilkAndy sorry I’ve been holding off - next has been broken for a little while. Should be good this afternoon.

**Comment by WilkAndy** [comment]
OK, thanks @loganharbour

**Comment by WilkAndy** [comment]
> Hi @WilkAndy and @cpgr, Many thanks for this!! This is great! Looking at the new example (multiblock.i) file you’ve added (Andy), and thinking about PorousFlowMaterialJoiner, I am wondering if we can safely assume that variables temp (in blocks 0 1; controlled by PorousFlow) and temp1 (in block 2; controlled by a non-PorousFlow Diffusion Kernel) in the example will be able to be coupled through their common bounding side?

No, @garciapintado , `temp` and `temp1` will not be coupled at all.  MOOSE will treat them as completely independent, since they don't both appear in any Kernel, or BC, or anything else.  The variable `pressure` will be defined over the whole mesh, and is governed by different physics in different parts of the mesh.  Probably that's the kind of thing you want.

**Comment by garciapintado** [comment]
> > Hi @WilkAndy and @cpgr, Many thanks for this!! This is great! Looking at the new example (multiblock.i) file you’ve added (Andy), and thinking about PorousFlowMaterialJoiner, I am wondering if we can safely assume that variables temp (in blocks 0 1; controlled by PorousFlow) and temp1 (in block 2; controlled by a non-PorousFlow Diffusion Kernel) in the example will be able to be coupled through their common bounding side?
> 
> No, @garciapintado , `temp` and `temp1` will not be coupled at all. MOOSE will treat them as completely independent, since they don't both appear in any Kernel, or BC, or anything else. The variable `pressure` will be defined over the whole mesh, and is governed by different physics in different parts of the mesh. Probably that's the kind of thing you want.

Ah, right! Many thanks.

**Review by loganharbour — APPROVED** [review]

**Pull request merged**